src/solver.rs

```diff
@@ -108,7 +49,67 @@ impl PackageResolutionStatistics {
 }
 
 /// Main function of the library.
+/// 
 /// Finds a set of packages satisfying dependency bounds for a given package + version pair.
+///
+/// It consists in efficiently finding a set of packages and versions
+/// that satisfy all the constraints of a given project dependencies.
+/// In addition, when that is not possible,
+/// PubGrub tries to provide a very human-readable and clear
+/// explanation as to why that failed.
+/// Below is an example of explanation present in
+/// the introductory blog post about PubGrub
+/// (Although this crate is not yet capable of building formatting quite this nice.)
+///
+/// ```txt
+/// Because dropdown >=2.0.0 depends on icons >=2.0.0 and
+///   root depends on icons <2.0.0, dropdown >=2.0.0 is forbidden.
+///
+/// And because menu >=1.1.0 depends on dropdown >=2.0.0,
+///   menu >=1.1.0 is forbidden.
+///
+/// And because menu <1.1.0 depends on dropdown >=1.0.0 <2.0.0
+///   which depends on intl <4.0.0, every version of menu
+///   requires intl <4.0.0.
+///
+/// So, because root depends on both menu >=1.0.0 and intl >=5.0.0,
+///   version solving failed.
+/// ```
+///
+/// TODO: there is not Version traits
```

Comment:
This paragraph is out of date and needs to be rewritten. I'm open to suggestions for how to explain `VS`.

Reply:
I'd link people to `DependencyProvider` as main interface, otherwise we have to repeat all of that information.